Sources/StreamChatUI/ChatMessageList/ChatMessageListView.swift

```diff
@@ -9,10 +9,52 @@ import UIKit
 open class ChatMessageListView: UITableView, Customizable, ComponentsProvider {
     private var identifiers: Set<String> = .init()
     private var isInitialized: Bool = false
-    /// Component responsible to process an array of `[ListChange<Item>]`'s and apply those changes to a view.
-    private lazy var listChangeUpdater: ListChangeUpdater = TableViewListChangeUpdater(
-        tableView: self
-    )
+
+    // MARK: - Difference Kit Snapshot Handling
+
+    // The internal properties below is to handle the DifferenceKit API. Currently it is
+    // internal because these should actually be handled in the `ChatMessageListVC` but
+    // that would make this class obsolete especially the `updateMessages(changes:)` and
+    // it would require a lot of breaking changes. So for now, the Diff logic will live here.
+
+    /// When inserting messages at the bottom, if the user is scrolled up,
+    /// we skip adding the message to the UI until the user scrolls back
+    /// to the bottom. This is to avoid message list jumps.
+    internal var skippedMessages: Set<MessageId> = [] {
+        didSet {
+            if skippedMessages.isEmpty {
+                newMessagesSnapshot = _newMessagesSnapshotWithSkipped
+            }
+        }
+    }
+
+    /// The previous messages snapshot before the next update.
+    internal var previousMessagesSnapshot: [ChatMessage] = [] {
+        didSet {
+            let isNotSkippedMessage: (ChatMessage) -> Bool = {
+                !self.skippedMessages.contains($0.id)
+            }
+            previousMessagesSnapshot = previousMessagesSnapshot.filter(isNotSkippedMessage)
+        }
+    }
+
+    /// The new messages snapshot reported by the channel or message controller.
+    private var _newMessagesSnapshotWithSkipped: [ChatMessage] = []
```

Comment:
Yeah, the more I see skipped messages related logic, the more concerned I get 🙏

Reply:
This is more complex than it actually should be because we are trying to avoid breaking changes. Otherwise, this would look very simple. This [commit](https://github.com/GetStream/stream-chat-swift/pull/2226/commits/955802ea856f903089547ec88674c71d7b2ad87e) shows all the skipped message logic before moving the code to the `ChatMessageListView`. Moving the logic to the message list view, in order to make it work for both the Channel List and Thread, makes things a little bit more tricky.

Reply:
I've tried to improve a little bit the readability as much as I could with the limited constraints we have, but let me know what you think !